projects/motor_controller_interface/src/motor_controller_fsm.c

```diff
@@ -0,0 +1,322 @@
+#include "can_unpack.h"
+#include "config.h"
+#include "exported_enums.h"
+#include "fifo.h"
+#include "motor_controller.h"
+#include "motor_controller_interface_events.h"
+
+static FSM s_fsm;
+static Fifo *s_fifo;
+static MotorControllerMeasurement *s_mc_measurements;
+static CanInterval *s_left_interval = NULL;
+static CanInterval *s_right_interval = NULL;
+
+// Generic "dumb" handler to copy telemetry values to a buffer
+static void prv_handle_slave_msg(const GenericCanMsg *msg, void *context) {
+  memcpy(context, &msg->data, MIN(msg->dlc, (size_t)8));
+}
+
+static StatusCode prv_push_drive(const CANMessage *msg, void *context, CANAckStatus *ack_reply) {
+  DriverControlsData data = { 0 };
+
+  CAN_UNPACK_MOTOR_CONTROLS(msg, &data.throttle, &data.direction, &data.cruise_control,
+                            &data.brake_state);
+
+  status_ok_or_return(fifo_push(s_fifo, &data));
+
+  return event_raise(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, 0);
+}
+
+static bool prv_mechanical_braking_guard(const FSM *fsm, const Event *e, void *context) {
+  DriverControlsData data = { 0 };
+
+  fifo_peek(s_fifo, &data);
+  if (data.brake_state != DRIVER_CONTROLS_BRAKE_DISENGAGED) {
+    // The pop will happen in the output function
+    return true;
+  }
+  return false;
+}
+
+static bool prv_torque_control_forward_guard(const FSM *fsm, const Event *e, void *context) {
+  // We can only go forward if the direction selector is set to Forward and the
+  // Current is a valid value
+  DriverControlsData data = { 0 };
+
+  fifo_peek(s_fifo, &data);
+  if (data.brake_state == DRIVER_CONTROLS_BRAKE_DISENGAGED &&
+      data.direction == DRIVER_CONTROLS_FORWARD && data.cruise_control == 0) {
+    // The pop will happen in the output function
+    return true;
+  }
+  return false;
+}
+
+static bool prv_torque_control_reverse_guard(const FSM *fsm, const Event *e, void *context) {
+  // We can only go reverse if the direction selector is set to Reverse and the
+  // Current is a valid value, and cruise control is disabled
+  DriverControlsData data = { 0 };
+
+  fifo_peek(s_fifo, &data);
+  if (data.brake_state == DRIVER_CONTROLS_BRAKE_DISENGAGED &&
+      data.direction == DRIVER_CONTROLS_REVERSE && data.cruise_control == 0) {
+    // The pop will happen in the output function
+    return true;
+  }
+  return false;
+}
+
+static bool prv_cruise_control_forward_guard(const FSM *fsm, const Event *e, void *context) {
+  // We can only go Cruise Control if the direction selector is set to Forward
+  // and the Current is a valid value
+  DriverControlsData data = { 0 };
+
+  fifo_peek(s_fifo, &data);
+  if (data.brake_state == DRIVER_CONTROLS_BRAKE_DISENGAGED &&
+      data.direction == DRIVER_CONTROLS_FORWARD && data.cruise_control != 0) {
+    // The pop will happen in the output function
+    return true;
+  }
+  return false;
+}
+
+// Motor Controller state definitions
+
+// Note: Cruise Control is invalid in reverse
+FSM_DECLARE_STATE(torque_control_forward);
```

Comment:
It's not officially stated, but we seem to have a convention of prefixing states with `state_`.